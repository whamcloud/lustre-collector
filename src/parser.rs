// Copyright (c) 2018 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::{mds::mds_parser, mgs::mgs_parser, oss::oss_parser, top_level_parser, types::Record};
use combine::{choice, error::ParseError, many, Parser, Stream};

pub fn params() -> Vec<String> {
    let mut a = top_level_parser::top_level_params();
    a.extend(mgs_parser::params());
    a.extend(oss_parser::params());
    a.extend(mds_parser::params());
    a
}

pub fn parse<I>() -> impl Parser<I, Output = Vec<Record>>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    many(choice((
        top_level_parser::parse(),
        mgs_parser::parse(),
        mds_parser::parse(),
        oss_parser::parse(),
    )))
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_params() {
        assert_debug_snapshot!(params());
    }
}
