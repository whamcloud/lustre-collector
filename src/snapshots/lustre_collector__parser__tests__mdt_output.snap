---
source: src/parser.rs
expression: result
---
(
    [
        Host(
            Memused(
                HostStat {
                    param: Param(
                        "memused",
                    ),
                    value: 343719411,
                },
            ),
        ),
        Host(
            MemusedMax(
                HostStat {
                    param: Param(
                        "memused_max",
                    ),
                    value: 344830779,
                },
            ),
        ),
        Host(
            LNetMemUsed(
                HostStat {
                    param: Param(
                        "lnet_memused",
                    ),
                    value: 140274209,
                },
            ),
        ),
        Host(
            HealthCheck(
                HostStat {
                    param: Param(
                        "health_check",
                    ),
                    value: "healthy",
                },
            ),
        ),
        Target(
            ConnectedClients(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "connected_clients",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 2,
                },
            ),
        ),
        Target(
            ConnectedClients(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "connected_clients",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 2,
                },
            ),
        ),
        Target(
            ContendedLocks(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "contended_locks",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 32,
                },
            ),
        ),
        Target(
            ContendedLocks(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "contended_locks",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 32,
                },
            ),
        ),
        Target(
            ContentionSeconds(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "contention_seconds",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 2,
                },
            ),
        ),
        Target(
            ContentionSeconds(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "contention_seconds",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 2,
                },
            ),
        ),
        Target(
            CtimeAgeLimit(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "ctime_age_limit",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 10,
                },
            ),
        ),
        Target(
            CtimeAgeLimit(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "ctime_age_limit",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 10,
                },
            ),
        ),
        Target(
            EarlyLockCancel(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "early_lock_cancel",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            EarlyLockCancel(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "early_lock_cancel",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockCount(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "lock_count",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockCount(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "lock_count",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockTimeouts(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "lock_timeouts",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockTimeouts(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "lock_timeouts",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockUnusedCount(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "lock_unused_count",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockUnusedCount(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "lock_unused_count",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LruMaxAge(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "lru_max_age",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 3900000,
                },
            ),
        ),
        Target(
            LruMaxAge(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "lru_max_age",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 3900000,
                },
            ),
        ),
        Target(
            LruSize(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "lru_size",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 800,
                },
            ),
        ),
        Target(
            LruSize(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "lru_size",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 800,
                },
            ),
        ),
        Target(
            MaxNolockBytes(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "max_nolock_bytes",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            MaxNolockBytes(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "max_nolock_bytes",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            MaxParallelAst(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "max_parallel_ast",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 1024,
                },
            ),
        ),
        Target(
            MaxParallelAst(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "max_parallel_ast",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 1024,
                },
            ),
        ),
        Target(
            ResourceCount(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "resource_count",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            ResourceCount(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "resource_count",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            Stats(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "md_stats",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: [
                        Stat {
                            name: "getattr",
                            units: "reqs",
                            samples: 4,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "statfs",
                            units: "reqs",
                            samples: 2,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                    ],
                },
            ),
        ),
        Target(
            Stats(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "md_stats",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: [
                        Stat {
                            name: "getattr",
                            units: "reqs",
                            samples: 2,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "statfs",
                            units: "reqs",
                            samples: 2,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                    ],
                },
            ),
        ),
        Target(
            NumExports(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "num_exports",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 15,
                },
            ),
        ),
        Target(
            NumExports(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "num_exports",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 7,
                },
            ),
        ),
        Target(
            FilesFree(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "filesfree",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 2621151,
                },
            ),
        ),
        Target(
            FilesFree(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "filesfree",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 2621167,
                },
            ),
        ),
        Target(
            FilesTotal(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "filestotal",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 2621440,
                },
            ),
        ),
        Target(
            FilesTotal(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "filestotal",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 2621440,
                },
            ),
        ),
        Target(
            BytesAvail(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "kbytesavail",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 1568641024,
                },
            ),
        ),
        Target(
            BytesAvail(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "kbytesavail",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 1568837632,
                },
            ),
        ),
        Target(
            BytesFree(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "kbytesfree",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 1836306432,
                },
            ),
        ),
        Target(
            BytesFree(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "kbytesfree",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 1836503040,
                },
            ),
        ),
        Target(
            BytesTotal(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "kbytestotal",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 1863647232,
                },
            ),
        ),
        Target(
            BytesTotal(
                TargetStat {
                    kind: MDT,
                    param: Param(
                        "kbytestotal",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 1863647232,
                },
            ),
        ),
    ],
    "",
)
